iscservice/demo/util.c

```diff
@@ -0,0 +1,295 @@
+/* -*- C -*- */
+/*
+ * Copyright (c) 2021 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ */
+
+#include <stdio.h>
+
+#include "lib/trace.h"        /* m0_trace_set_mmapped_buffer */
+#include "rpc/rpclib.h"       /* M0_RPCLIB_MAX_RETRIES */
+#include "motr/client_internal.h" /* m0_client */
+#include "layout/plan.h"      /* m0_layout_io_plop */
+
+#include "iscservice/isc.h"
+#include "util.h"
+
+#ifndef DEBUG
+#define DEBUG 0
+#endif
+
+/* static variables */
+static struct m0_container      container;
+static struct m0_idx_dix_config dix_conf = {};
+
+/* global variables */
+struct m0_realm     uber_realm;
+int                 trace_level = 0;
+bool                m0trace_on = false;
+struct m0_semaphore isc_sem;
+struct m0_list      isc_reqs;
+
+/**
+ * Return parity group size for object.
+ */
+uint64_t isc_m0gs(struct m0_obj *obj, struct m0_client *cinst)
+{
+	unsigned long           usz; /* unit size */
+	struct m0_pool_version *pver;
+
+	pver = m0_pool_version_find(&cinst->m0c_pools_common,
+				    &obj->ob_attr.oa_pver);
+	if (pver == NULL) {
+		ERR("invalid object pool version: "FID_F"\n",
+		    FID_P(&obj->ob_attr.oa_pver));
+		return 0;
+	}
+
+	usz = m0_obj_layout_id_to_unit_size(obj->ob_attr.oa_layout_id);
+
+	return usz * pver->pv_attr.pa_N;
+}
+
+void free_segs(struct m0_bufvec *data, struct m0_indexvec *ext,
+	       struct m0_bufvec *attr)
+{
+	m0_indexvec_free(ext);
+	m0_bufvec_free(data);
+	m0_bufvec_free(attr);
+}
+
+int alloc_segs(struct m0_bufvec *data, struct m0_indexvec *ext,
+	       struct m0_bufvec *attr, uint64_t bsz, uint32_t cnt)
+{
+	int i, rc;
+
+	rc = m0_bufvec_alloc(data, cnt, bsz) ?:
+	     m0_bufvec_alloc(attr, cnt, 1) ?:
+	     m0_indexvec_alloc(ext, cnt);
+	if (rc != 0)
+		goto err;
+
+	for (i = 0; i < cnt; i++)
+		attr->ov_vec.v_count[i] = 0; /* no attrs */
+
+	return 0;
+ err:
+	free_segs(data, ext, attr);
+	return rc;
+}
+
+uint64_t set_exts(struct m0_indexvec *ext, uint64_t off, uint64_t bsz)
+{
+	uint32_t i;
+
+	for (i = 0; i < ext->iv_vec.v_nr; i++) {
+		ext->iv_index[i] = off;
+		ext->iv_vec.v_count[i] = bsz;
+		off += bsz;
+	}
+
+	return i * bsz;
+}
+
+int isc_req_prepare(struct isc_req *req, struct m0_buf *args,
```

Comment:
This is a common function, and can be used later for other ISC?

As an example, it's better to document all the parameters.